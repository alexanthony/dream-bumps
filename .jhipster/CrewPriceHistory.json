{
    "fluentMethods": true,
    "relationships": [
        {
            "relationshipName": "crew",
            "otherEntityName": "crew",
            "relationshipType": "many-to-one",
            "relationshipValidateRules": [
                "required"
            ],
            "otherEntityField": "id"
        }
    ],
    "fields": [
        {
            "fieldName": "dateTime",
            "fieldType": "ZonedDateTime",
            "fieldValidateRules": [
                "required"
            ]
        },
        {
            "fieldName": "price",
            "fieldType": "BigDecimal",
            "fieldValidateRules": [
                "required",
                "min"
            ],
            "fieldValidateRulesMin": 0
        }
    ],
    "changelogDate": "20170305151406",
    "dto": "no",
    "service": "no",
    "entityTableName": "crew_price_history",
    "pagination": "no"
}
