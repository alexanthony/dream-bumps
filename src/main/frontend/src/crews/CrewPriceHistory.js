import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import {Card, CardTitle, CardText} from 'material-ui/Card';
import { Line } from 'react-chartjs-2';

import { fetchCrewPriceHistory } from './actions';
import { getCrewPriceHistory } from '../rootReducer';

const minPrice = priceHistories => priceHistories.reduce((prev, curr) => (prev < curr.price ? prev : curr.price), Number.POSITIVE_INFINITY);

const maxPrice = priceHistories => priceHistories.reduce((prev, curr) => (prev > curr.price ? prev : curr.price), 0);

/**
 * CrewPriceHistory
 */
class CrewPriceHistory extends Component {
  componentWillMount() {
    if (this.props.crewId) {
      this.props.fetchCrewPriceHistory(this.props.crewId);
    }
  }
  render() {
    return (
      <Card>
        <CardTitle title="Price History" />
        <CardText>
          <Line
            data={{datasets: [{data: this.props.crewPriceHistory.map(history => ({x: history.dateTime, y: history.price})),
                      fill: false,
                      steppedLine: true,}]}}
            options={{
                scales: {
                  xAxes: [{type:'time'}],
                  yAxes: [{
                    ticks: {
                      suggestedMin: minPrice(this.props.crewPriceHistory) - 10,
                      suggestedMax: maxPrice(this.props.crewPriceHistory) + 10
                  }}]
                },
                legend: {
                  display: false
                }
              }}
          />
        </CardText>
      </Card>
    );
  }
}

CrewPriceHistory.propTypes = {
  crewPriceHistory: PropTypes.array,
  crewId: PropTypes.string,
  fetchCrewPriceHistory: PropTypes.func
};

export const mapStateToProps = (state, ownProps) => ({
  crewPriceHistory: getCrewPriceHistory(state)(ownProps.crewId)
});

export {CrewPriceHistory};

export default connect(mapStateToProps, { fetchCrewPriceHistory })(CrewPriceHistory);
