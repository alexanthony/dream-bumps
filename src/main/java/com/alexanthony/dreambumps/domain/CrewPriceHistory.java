package com.alexanthony.dreambumps.domain;

import org.hibernate.annotations.Cache;
import org.hibernate.annotations.CacheConcurrencyStrategy;

import javax.persistence.*;
import javax.validation.constraints.*;
import java.io.Serializable;
import java.math.BigDecimal;
import java.time.ZonedDateTime;
import java.util.Objects;

/**
 * A CrewPriceHistory.
 */
@Entity
@Table(name = "crew_price_history")
@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
public class CrewPriceHistory implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    @Column(name = "date_time", nullable = false)
    private ZonedDateTime dateTime;

    @NotNull
    @DecimalMin(value = "0")
    @Column(name = "price", precision=10, scale=2, nullable = false)
    private BigDecimal price;

    @ManyToOne(optional = false)
    @NotNull
    private Crew crew;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public ZonedDateTime getDateTime() {
        return dateTime;
    }

    public CrewPriceHistory dateTime(ZonedDateTime dateTime) {
        this.dateTime = dateTime;
        return this;
    }

    public void setDateTime(ZonedDateTime dateTime) {
        this.dateTime = dateTime;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public CrewPriceHistory price(BigDecimal price) {
        this.price = price;
        return this;
    }

    public void setPrice(BigDecimal price) {
        this.price = price;
    }

    public Crew getCrew() {
        return crew;
    }

    public CrewPriceHistory crew(Crew crew) {
        this.crew = crew;
        return this;
    }

    public void setCrew(Crew crew) {
        this.crew = crew;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        CrewPriceHistory crewPriceHistory = (CrewPriceHistory) o;
        if (crewPriceHistory.id == null || id == null) {
            return false;
        }
        return Objects.equals(id, crewPriceHistory.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "CrewPriceHistory{" +
            "id=" + id +
            ", dateTime='" + dateTime + "'" +
            ", price='" + price + "'" +
            '}';
    }
}
