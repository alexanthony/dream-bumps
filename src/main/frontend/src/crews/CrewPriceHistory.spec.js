import React from 'react';
import { shallow } from 'enzyme';
import { CrewPriceHistory } from './CrewPriceHistory';
import toJson from 'enzyme-to-json';

test('Renders properly', () => {
  const fetchCrewPriceHistory = jest.fn();
  const crewPriceHistory = [
    {price: 100, dateTime: '2017-05-02T21:51:15.123Z'},
    {price: 102, dateTime: '2017-05-02T21:53:15.123Z'},
    {price: 99, dateTime: '2017-05-04T12:51:15.123Z'}
  ];
  const component = shallow(
    <CrewPriceHistory
      crewPriceHistory={crewPriceHistory}
      crewId="2"
      fetchCrewPriceHistory={fetchCrewPriceHistory}
    />
  );
  expect(component).toBeDefined();
  expect(toJson(component)).toMatchSnapshot()
  expect(fetchCrewPriceHistory).toHaveBeenCalled();
  expect(fetchCrewPriceHistory.mock.calls.length).toBe(1);
  expect(fetchCrewPriceHistory.mock.calls[0][0]).toBe('2');
});
