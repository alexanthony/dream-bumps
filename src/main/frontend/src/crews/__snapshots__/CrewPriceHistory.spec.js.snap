exports[`test Renders properly 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}>
  <CardTitle
    title="Price History" />
  <CardText>
    <Line
      data={
        Object {
          "datasets": Array [
            Object {
              "data": Array [
                Object {
                  "x": "2017-05-02T21:51:15.123Z",
                  "y": 100,
                },
                Object {
                  "x": "2017-05-02T21:53:15.123Z",
                  "y": 102,
                },
                Object {
                  "x": "2017-05-04T12:51:15.123Z",
                  "y": 99,
                },
              ],
              "fill": false,
              "steppedLine": true,
            },
          ],
        }
      }
      options={
        Object {
          "legend": Object {
            "display": false,
          },
          "scales": Object {
            "xAxes": Array [
              Object {
                "type": "time",
              },
            ],
            "yAxes": Array [
              Object {
                "ticks": Object {
                  "suggestedMax": 112,
                  "suggestedMin": 89,
                },
              },
            ],
          },
        }
      } />
  </CardText>
</Card>
`;
